sources/prairiedog/src/main.rs

```diff
@@ -0,0 +1,359 @@
+    let state = State::load().context(error::LoadState)?;
+    let boot_partition_path = &state.active_set().boot;
+    let flags = nix::mount::MsFlags::MS_RELATIME
+        | nix::mount::MsFlags::MS_NOSUID
+        | nix::mount::MsFlags::MS_NOATIME;
+
+    info!(
+        "Mounting {} in {}",
+        boot_partition_path.display(),
+        BOOT_MOUNT_PATH
+    );
+    // Mount the active boot partition in /boot
+    nix::mount::mount(
+        Some(boot_partition_path),
+        BOOT_MOUNT_PATH,
+        Some("ext4"),
+        nix::mount::MsFlags::MS_RDONLY | flags,
+        Some(BOOT_MOUNT_SELINUX_CONTEXT),
+    )
+    .context(error::Mount {
+        path: BOOT_MOUNT_PATH,
+    })?;
+
+    // Make the mount point private so new mount namespaces don't have
+    // access to it. This has to be set as a different call otherwise
+    // the mount syscall returns `EINVAL`
+    nix::mount::mount(
+        NONE,
+        BOOT_MOUNT_PATH,
+        NONE,
+        nix::mount::MsFlags::MS_PRIVATE,
+        NONE,
+    )
+    .context(error::SetupMount {
+        path: BOOT_MOUNT_PATH,
+    })?;
+
+    Ok(())
+}
+
+/// Loads the crash kernel using kexec-tools
+fn load_crash_kernel() -> Result<()> {
+    let kexec_crash_size_path = Path::new(KEXEC_CRASH_SIZE);
+    let kexec_crash_size = fs::read(kexec_crash_size_path).context(error::ReadFile {
+        path: kexec_crash_size_path,
+    })?;
+    let memory_allocated = String::from_utf8_lossy(&kexec_crash_size);
+
+    // We provide a more useful message when no memory was reserved for the crash kernel. Exit
+    // gracefully since the user could have decided to use a tiny host, and the system shouldn't be
+    // in "degraded" state
+    if memory_allocated.trim() == "0" {
+        info!("No memory assigned for crash kernel. If you want to use kdump, please make sure the host has at least 2GB of memory");
+        return Ok(());
+    }
+
+    let kexec_load_disabled_path = Path::new(KEXEC_LOAD_DISABLED);
+    let kexec_load_disabled_value =
+        fs::read(kexec_load_disabled_path).context(error::ReadFile {
+            path: kexec_load_disabled_path,
+        })?;
+    let kexec_load_disabled = String::from_utf8_lossy(&kexec_load_disabled_value).trim() == "1";
+
+    // We provide a more useful message when `kexec_load_disabled` is set to 1
+    if kexec_load_disabled {
+        return error::KexecLoadDisabled.fail();
+    }
+
+    // Conditionally add `irqpoll` depending on the architecture
+    let kexec_cmd_line = if cfg!(target_arch = "x86_64") {
+        String::from(KEXEC_CMD_LINE) + " irqpoll"
+    } else {
+        String::from(KEXEC_CMD_LINE)
+    };
+
+    info!("Loading crash kernel");
+    // Load the panic kernel from `BOOT_MOUNT_PATH`, letting kexec decide which syscall
+    // it should it use (KEXEC_LOAD and KEXEC_FILE_LOAD). We will use this setting until
+    // we figure out why kexec doesn't recognize the 5.4 kernel image as a valid file
+    // to be used with KEXEC_FILE_LOAD.
```

Comment:
I got to port the kernel patch to make the `kexec_file_load` syscall work in the x86_64 5.4 kernel. However, I'm still waiting for the kernel folks to help me out with porting the syscall for aarch64, since the syscall was introduced for aarch64 on a higher kernel version.
 
I talked to @tjkirch since he expressed his concerns of using different syscalls depending on the kernel version/architecture. We agreed that we can proceed with my PR provided that we document in an GH issue that we will force the `kexec_file_load` syscall once the kernel folks add support for it for the 5.4 kernel.

Reply:
I think we should document the restrictions in the same place we document the feature, not (only) in a GH issue.